File: typescript_test_crate/tests/10_state_var_basic.ts
class StateVarBasic {
    private _state: string;
    private _state_stack: Array<any>;
    private _state_context: Record<string, any>;
    private _return_value: any;

    constructor() {
        this._state_stack = [];
        this._state_context = {  };
        this._return_value = null;
        this._state = "Counter";
        this._enter();
    }

    private _transition(target_state: string, exit_args: any = null, enter_args: any = null) {
        if (exit_args) {
            this._exit(...exit_args);
        } else {
            this._exit();
        }
        this._state = target_state;
        if (enter_args) {
            this._enter(...enter_args);
        } else {
            this._enter();
        }
    }

    private _change_state(target_state: string) {
        this._state = target_state;
    }

    private _dispatch_event(event: string, ...args: any[]) {
        const handler_name = `_s_${this._state}_${event}`;
        const handler = (this as any)[handler_name];
        if (handler) {
            return handler.apply(this, args);
        }
    }

    private _enter(...args: any[]) {
        if (this._state === "Counter") {
            this._state_context["count"] = 0;
        }
    }

    private _exit(...args: any[]) {
        // No exit handlers
    }

    public increment(): number {
        this._return_value = null
        this._dispatch_event("increment")
        return this._return_value
    }

    public get_count(): number {
        this._return_value = null
        this._dispatch_event("get_count")
        return this._return_value
    }

    public reset() {
        this._dispatch_event("reset");
    }

    private _s_Counter_get_count() {
        this._return_value = this._state_context["count"];
        return this._return_value;;
    }

    private _s_Counter_reset() {
        this._state_context["count"] = 0;
    }

    private _s_Counter_increment() {
        this._state_context["count"] = this._state_context["count"] + 1;
        this._return_value = this._state_context["count"];
        return this._return_value;;
    }
}


function main() {
    console.log("=== Test 10: State Variable Basic ===");
    const s = new StateVarBasic();

    // Initial value should be 0
    if (s.get_count() !== 0) {
        throw new Error(`Expected 0, got ${s.get_count()}`);
    }
    console.log(`Initial count: ${s.get_count()}`);

    // Increment should return new value
    let result = s.increment();
    if (result !== 1) {
        throw new Error(`Expected 1 after first increment, got ${result}`);
    }
    console.log(`After first increment: ${result}`);

    // Second increment
    result = s.increment();
    if (result !== 2) {
        throw new Error(`Expected 2 after second increment, got ${result}`);
    }
    console.log(`After second increment: ${result}`);

    // Reset should set back to 0
    s.reset();
    if (s.get_count() !== 0) {
        throw new Error(`Expected 0 after reset, got ${s.get_count()}`);
    }
    console.log(`After reset: ${s.get_count()}`);

    console.log("PASS: State variable basic operations work correctly");
}

main();

File: python_test_crate/tests/10_state_var_basic.py
from typing import Any, Optional, List, Dict, Callable

class StateVarBasic:
    def __init__(self):
        self._state_stack = []
        self._state_context = {}
        self._return_value = None
        self._state = "Counter"
        self._enter()

    def _transition(self, target_state, exit_args = None, enter_args = None):
        if exit_args:
            self._exit(*exit_args)
        else:
            self._exit()
        self._state = target_state
        if enter_args:
            self._enter(*enter_args)
        else:
            self._enter()

    def _change_state(self, target_state):
        self._state = target_state

    def _dispatch_event(self, event, *args):
        handler_name = f"_s_{self._state}_{event}"
        handler = getattr(self, handler_name, None)
        if handler:
            return handler(*args)

    def _enter(self, *args):
        if self._state == "Counter":
            self._state_context["count"] = 0


    def _exit(self, *args):
        # No exit handlers
        pass

    def increment(self) -> int:
        self._return_value = None
        self._dispatch_event("increment")
        return self._return_value

    def get_count(self) -> int:
        self._return_value = None
        self._dispatch_event("get_count")
        return self._return_value

    def reset(self):
        self._dispatch_event("reset")

    def _s_Counter_get_count(self) -> int:
        self._return_value = self._state_context["count"]
        return

    def _s_Counter_reset(self):
        self._state_context["count"] = 0

    def _s_Counter_increment(self) -> int:
        self._state_context["count"] = self._state_context["count"] + 1
        self._return_value = self._state_context["count"]
        return


def main():
    print("=== Test 10: State Variable Basic ===")
    s = StateVarBasic()

    # Initial value should be 0
    assert s.get_count() == 0, f"Expected 0, got {s.get_count()}"
    print(f"Initial count: {s.get_count()}")

    # Increment should return new value
    result = s.increment()
    assert result == 1, f"Expected 1 after first increment, got {result}"
    print(f"After first increment: {result}")

    # Second increment
    result = s.increment()
    assert result == 2, f"Expected 2 after second increment, got {result}"
    print(f"After second increment: {result}")

    # Reset should set back to 0
    s.reset()
    assert s.get_count() == 0, f"Expected 0 after reset, got {s.get_count()}"
    print(f"After reset: {s.get_count()}")

    print("PASS: State variable basic operations work correctly")

if __name__ == '__main__':
    main()

